src/op.rs

```diff
@@ -740,12 +711,25 @@ where
 }
 
 #[derive(Debug)]
-enum EvaluationState<R: Reader> {
+enum EvaluationState {
     Start(Option<u64>),
     Ready,
     Error(Error),
     Complete,
-    Waiting(OperationEvaluationResult<R>),
+    Waiting(EvaluationWaiting),
+}
+
+#[derive(Debug)]
+enum EvaluationWaiting {
+    Memory,
+    Register { offset: i64 },
+    FrameBase { offset: i64 },
+    Tls,
+    Cfa,
+    AtLocation,
+    EntryValue,
+    ParameterRef,
+    TextBase { offset: u64 },
```

Comment:
Since @khuey introduced the waiting stuff so that he could integrate well with tokio/futures APIs, it seems to me like it would be good to get his take on this commit's changes.

Looks good to me, FWIW.

Reply:
As far as I can tell this doesn't change the public API at all so it's fine for me. I don't really see why this is a win *internally* though. If I understand what this does correctly it violates the "make impossible states unrepresentable" principle by allowing an `OperationEvaluationResult` to contain mismatched `EvaluationWaiting` and `EvaluationResult` enum values.

Reply:
I don't understand how this is worse than before in terms of "make impossible states unrepresentable". There was nothing before that disallowed a mismatch between the state and the result either (ie the value stored in `EvaluationState::state` and the value returned from `Evaluation::evaluate`).

It's a win IMO because the logic for handling an operation is all in one spot, instead of split between `evaluate_one_operation` and `evaluate_internal`.

Reply:
Mmm I guess that's fair.

Carry on.